Replace existing docker networking code and integrate with libnetwork

mrjana opened the pull request:
This PR attempts to integrate libnetwork with three main goals:
- Modularize docker networking code to a separate package
- Introduce the first implementation of Container Network Model and make docker purely rely on that
- Allow more kinds of network drivers to integrated with ease and possibly no change to docker core at all

Please refer to following libnetwork docs for more info:
- [README](https://github.com/docker/libnetwork/blob/master/README.md)
- [ROADMAP](https://github.com/docker/libnetwork/blob/master/ROADMAP.md)
- [Design](https://github.com/docker/libnetwork/blob/master/docs/design.md)

All dead code has been removed.

mrjana committed:
Docker integration with libnetwork

    - Updated Dockerfile to satisfy libnetwork GOPATH requirements.
    - Reworked daemon to allocate network resources using libnetwork.
    - Reworked remove link code to also update network resources in libnetwork.
    - Adjusted the exec driver command population to reflect libnetwork design.
    - Adjusted the exec driver create command steps.
    - Updated a few test cases to reflect the change in design.
    - Removed the dns setup code from docker as resolv.conf is entirely managed
      in libnetwork.
    - Integrated with lxc exec driver.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

mrjana committed:
Vendoring in libnetwork, netns, netlink

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

crosbymichael commented:
you need to rebase ;)

mrjana commented:
@crosbymichael done :-)

jfrazelle commented:
started test on lxc ;) feeling brave

LK4D4 commented:
@mrjana Also, you should separate commit message body from title with empty line

calavera commented:
very nice at first glance, thanks for vendoring the code in a separated commit, much better to review :zap:

cpuguy83 commented:
Test failure.

mavenugo commented:
@cpuguy83 thanks. am looking into it.

mavenugo commented on hack/vendor.sh:
```diff
@@ -53,6 +54,10 @@ clone git github.com/kr/pty 05017fcccf
 clone hg code.google.com/p/go.net 84a4013f96e0
 clone hg code.google.com/p/gosqlite 74691fb6f837
+clone git github.com/go-fsnotify/fsnotify v1.2.0
+clone git github.com/go-check/check 64131543e7896d5bcc6bd5a76287eb75ea96c673
```
can you pls remove the redundant go-fsnotify and go-check dependencies ?

mrjana commented:
ping @crosbymichael @jfrazelle @LK4D4 @calavera @cpuguy83

jfrazelle commented:
> Please note that the code not needed in docker any more like the networkdriver and portmapper haven't been removed yet. It can be removed in subsequent PRs once this one gets merged.

So is that code dead or is it still being called.

jfrazelle commented:
as far as design goes I like it, it is clean and simple.

crosbymichael commented:
Yes, don't leave dead code around, it will just be confusing for anyone changing code after you.  Don't comment things out either, just delete, everything is in git so we can always see what it was before.

mrjana commented:
@crosbymichael Sure, let me cleanup dead code.

@jfrazelle I ran lxc integration tests in my dev machine and it had only 7 failures(which was the same on stock 1.7.0-dev in my machine). Does lxc test fail intermittently?

mrjana committed:
Removed dead code from docker after libnetwork integration

As part of this some generic packages like iptables, etchosts and resolvconf
have also been moved to libnetwork. Even though they can still be
consumed in a generic fashion they will reside and be maintained
from within the libnetwork project.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

mrjana commented:
@jfrazelle @crosbymichael @LK4D4 
Updated the PR with all the comments taken care

LK4D4 commented:
With last commit it is definetely better.

dave-tucker commented:
Just ran some manual smoke tests on this PR and everything seems good ... :+1:

Environment:
``` bash
$ make BIND_DIR=bundles binary
$ docker-machine create libnetwork
$ scp -i ~/.docker/machine/machines/libnetwork/id_rsa bundles/1.7.0-dev/binary/docker-1.7.0-dev docker@`docker-machine ip`:docker-1.7.0-dev
$ docker-machine ssh
docker@libnetwork:~$ sudo mv docker-1.7.0-dev /usr/local/bin
docker@libnetwork:~$ sudo mv /usr/local/bin/docker /usr/local/bin/docker.old
docker@libnetwork:~$ sudo ln -s /usr/local/bin/docker-1.7.0-dev /usr/local/bin/docker
docker@libnetwork:~$ sudo /etc/init.d/docker restart
docker@libnetwork:~$ exit
```

Tests:
```
# Create 50 containers in bridge mode
for i in $(seq 1 50); do docker run -itd --name foo$i busybox ; done
# Ping them all from foo1
for i in $(seq 1 50); do docker exec foo1 ping -c4 $(docker inspect --format "{{.NetworkSettings.IPAddress}}" foo$i); done
# Check they can all reach 8.8.8.8
for i in $(seq 1 50); do docker exec foo$i ping -c2 8.8.8.8; done
# Check apt-get
docker run -it --rm debian;jessie apt-get update
# Check port mapping
docker run -itd --name nginx -P nginx
port=$(docker port nginx | grep 80/tcp | sed s/.*://)
curl http://`docker-machine ip`:$port
# Check links
docker run -it --rm --link nginx:nginx debian:jessie
cat /etc/hosts | grep nginx
env | grep NGINX
exit
# Check --net=none
docker run -it --rm --net=none debian:jessie
ip addr show dev eth0 || echo "YAY"
# Check --net=host
docker run -itd --net=host --name=host debian:jessie
docker inspect --format {{.NetworkSettings.SandboxKey}} host | grep /var/run/netns/default
docker attach host
ifconfig
exit
# Check --net=container
docker run -it --rm --name=foo51 --net=container:foo50 busybox ifconfig | grep $(docker inspect --format {{.NetworkSettings.IPAddress}} foo50)
# Check Veth Cleanup
for x in $(docker ps -a -q) ; do docker stop $x; done
ip addr | grep veth || echo "YAY"
```

@mrjana I see the endpoint lifecycle isn't decoupled from the container yet.
Is that following in a subsequent PR?

```
## FAILS
# Check container IP persists on stop/start
foo1_ip=$(docker inspect --format {{.NetworkSettings.IPAddress}} foo1)
docker stop foo1
docker start foo1
docker exec foo1 ifconfig | grep $foo1_ip
```

mavenugo commented:
@dave-tucker Thanks for your help. 

> @mrjana I see the endpoint lifecycle isn't decoupled from the container yet. Is that following in a subsequent PR?

Yes. This is expected for now. We tried our best to keep the existing behavior in place with this PR. The endpoint life-cycle decoupling will follow in subsequent PR.

mrjana commented:
@dave-tucker To clarify a bit more, the endpoint lifecycle for implicitly created endpoints like the ones created when we invoke `docker run --net:<net_name>` without any endpoint name, will actually coincide with the container lifecycle itself. It is the endpoints which are explicitly created by the user which will have it's own lifecycle independent of the container.

Nevertheless, there is a strong need still to provide the same network settings like IP addresses etc. to a restarting container and we will achieve this using an endpoint caching logic inside libnetwork in such a way that we will retain a deleted endpoint in a cache inside libnetwork and not notify the driver and if a container comes back and tries creating the same endpoint (read same name) then we will try to give it back from the cache. Ofcourse, the cache ages out after some time so that we don't hold on to network resources indefinitely.

dave-tucker commented:
@mrjana @mavenugo thanks for clarifying!

codergr commented:
Thanks @mavenugo like the design

squaremo commented:
> we will achieve this using an endpoint caching logic inside libnetwork

This may not play so well with remote drivers, which in general need to know explicitly when resources are released (and may wish to do their own caching / expiration).

LK4D4 commented:
@mrjana need rebase.

mrjana commented:
@LK4D4 rebase done.

LK4D4 commented:
@mrjana Seems like you removed `fsnotify` usage. Can you remove it from vendor script and directory?

LK4D4 commented:
Looks like `daemon/networkdriver/bridge/driver.go` still here.

mrjana commented:
@LK4D4 @calavera Updated the PR with the comments taken care

estesp commented:
Per @LK4D4 comment re: fsnotify, why is the `/etc/resolv.conf` updating feature removed?  I notice that tests testing this functionality were left in the test suite but gutted from actually testing the feature.   That seems wrong--we need tests removed and docs updated if we are all agreed on removing an actual function from the current network stack's capabilities.

LK4D4 commented:
Yes, @estesp has good questions.

mrjana commented:
@estesp I was trying to get your comments on this in IRC but couldn't find you. Anyway the test case in integration-cli was testing the design while we should be testing the expected functionality. 

The expected function as I understand(correct me if I am wrong) is if the container got it's /etc/resolv.conf from the host(and not from daemon or container config option) and if the host entries changed then in the next restart of the container update the container's /etc/resolv.conf to reflect the host entries. The test case in integration-cli is still testing that functionality. But we changed the design of how this is achieved so if the test case was testing the design then it is not applicable any more. That's why removed the bits which were testing the old design which was testing if the container file got updated for  a change in host /etc/resolv.conf while the container is not running.

tiborvass commented:
Ping @jfrazelle @icecrime @duglin @estesp @unclejack @vishh @vieux @aluzzardi @dmp42 @dmcgowan @ehazlett @stevvooe @endophage 

Would really appreciate help reviewing this very important PR that would lay the foundation for future docker networking improvements.

If you're not familiar with networking, just make sure the refactoring makes sense.
If you feel familiar with networking, please be critical about what might break given this refactor.

Thanks in advance for your help

estesp commented:
@mrjana I'm "estesp" on IRC in docker, docker-{dev,maintainers,distribution}.  I can join -network if discussions should happen there, however, this afternoon I'm offline most of the rest of the day.

To your point about the test--it was actually testing function and edge cases around the updater, using the knowledge of the design as an easy validation method.  If the design changes, then maybe some other method for validation is required, but I don't think any of the tests are invalid unless the function has changed.  The variations were around - a) the user changing the container's resolv.conf manaully--which means no auto-updates should happen, b) running containers should not be updated, and c) the act of updating should retain the default filtering out of localhost resolvers.

If any of those are no longer valid, then we need to discuss why the design changed and what impact to docs or other expectations of users are.

cpuguy83 commented:
Made some general code comments, haven't tested thoroughly yet.

mrjana commented:
@estesp Fair point on testing the variations. The earlier tests were testing this assuming certain design. I will try reworking the test code to include testing these variations purely in a black box way.

calavera commented:
I'm not completely familiar with all the features that we're modifying here, but code LGTM over all. I agree that we should try to replace the removed tests with better tests :smile:

LK4D4 commented:
Yeah, actually I'm sure we need consensus between @mrjana and @estesp about `resolv.conf` feature.

LK4D4 commented:
@ashahab-altiscale Would you mind to look on lxc trick about setns.

ashahab-altiscale commented:
Sure. Btw, is there a design proposal?
On May 15, 2015 3:00 PM, "Alexander Morozov" <notifications@github.com>
wrote:

> @ashahab-altiscale <https://github.com/ashahab-altiscale> Would you mind
> to look on lxc trick about setns.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13060#issuecomment-102534535>.
>

mrjana commented:
@ashahab-altiscale For libnetwork design please take a look at the link in the PR description above and there is a link to the design. 

But fundamentally what happens is libnetwork manages all the network resources for the containers with the help of a set of drivers. It also manages the net namespace in which these network resources are going to be present. When it has plumbed all the network resources into the netns it will just pass along the netns to the container's exec driver to use when it sets up the execution environment. Both libcontainer and lxc exec drivers use this to setup the execution enviroment.

Let me know if you need any clarifications.

stevvooe commented on api/server/server.go:
```diff
@@ -1640,8 +1640,9 @@ func allocateDaemonPort(addr string) error {
 		return fmt.Errorf("failed to lookup %s address in host specification", host)
 	}
 
+	pa := portallocator.Get()
```
This is a very java-y pattern for Go. It would make more sense to have a `portallocator.Default` and set it up in the package `init` function.

cpuguy83 replied:
@stevvooe #11946

stevvooe replied:
Then hide the loading somewhere else. This getter pattern doesn't belong in Go code. Arguably, if file io is happening, this shouldn't be a global.

LK4D4 replied:
I agree with @stevvooe

mrjana replied:
@stevvooe We are hiding the loading behind the `Get`. So to clarify you seem to be opposed to naming this method `Get` If it's called Default() would that be ok? The file io that is happening is on `/proc/sys/net/ipv4/ip_local_port_range` which is a global read only file to read the local port range configured in the system. So this is supposed to be global despite that fact that file io is happening.

LK4D4 replied:
@stevvooe Proposing to have variable named `Default` and load on import.

stevvooe replied:
It might be better to follow the convention of the [flag](http://godoc.org/flag) package. In addition to having an "instance" creation function, like `NewAllocator`, promote all the same methods to the package level. For the instances that can't be created with `init`, such as those that do file io, call `sync.Do` at the start of each method to ensure it is created. It is a little more code on the package side by makes the pattern clear.

Let me know if I am not clear above.

mrjana replied:
@stevvooe. We do indeed use `sync.Once` for initializing the default instance. Just that we were only doing it once when `portallocator.New` was called. This way we don't need to do `once.Do` on all the methods. In fact this is how our original PR was, just naming this `New` but because of a previous comment we changed it to `Get`

stevvooe replied:
I would just follow the conventions of the flag package. Don't worry about the cost of checking `sync.Once`. It will not be a bottleneck.

mavenugo replied:
Thanks @stevvooe. The port allocator package needs a redesign to reflect few other issues as captured in (https://github.com/docker/libnetwork/issues/169). This PR tries to port the existing functionality without impacting the behavior so that we have apples-apples comparison. If you are okay, we can address your concern along with the above enhancement. Please let us know.

stevvooe replied:
SGTM!

stevvooe commented on daemon/config.go:
```diff
@@ -17,7 +17,7 @@ const (
 type CommonConfig struct {
 	AutoRestart    bool
-	Bridge         bridge.Config
+	Bridge         bridgeConfig
```
I know the surrounding fields are documented, but let's not continue that. Please document according to [this blog](http://blog.golang.org/godoc-documenting-go-code).

mrjana replied:
will fix.

stevvooe commented on daemon/config.go:
```diff
@@ -34,6 +34,24 @@ type CommonConfig struct {
+// configuration.
+type bridgeConfig struct {
+	EnableIPv6                  bool
+	EnableIptables              bool
```
`EnableIPTables`. [Initialism](https://github.com/golang/go/wiki/CodeReviewComments#initialisms).

mrjana replied:
Will fix it.

stevvooe commented on daemon/container.go:
```diff
@@ -559,19 +532,69 @@ func (container *Container) buildHostnameFile() error {
 
-	var extraContent []etchosts.Record
+	if len(container.hostConfig.Dns) > 0 {
+		dns = container.hostConfig.Dns
```
Not sure where this is defined but it should be `container.hostConfig.DNS`.

mrjana replied:
`docker/docker` code has a lot of these occurrences(infact this was not introduced by this PR) and  I don't think we can clean up all of that in this PR. So what I would suggest is that we should open a seperate PR and fix all of these and enable `golint` hook so that it catches these automatically. In libnetwork we do exactly that and we don't have any initialism violations.

stevvooe replied:
@mrjana We does this as well, over in distribution. I do like to fix them where we can. Please ignore if it doesn't make sense here.

stevvooe commented on daemon/container.go:
```diff
@@ -559,19 +532,69 @@ func (container *Container) buildHostnameFile() error {
+	if len(container.hostConfig.DnsSearch) > 0 {
+		dnsSearch = container.hostConfig.DnsSearch
+	} else if len(container.daemon.config.DnsSearch) > 0 {
+		dnsSearch = container.daemon.config.DnsSearch
```
`DNSSearch`.

mrjana replied:
Same comment as above.

stevvooe commented on daemon/container.go:
```diff
@@ -582,60 +605,226 @@ func (container *Container) buildHostsFiles(IP string) error {
+		}
+	}
+
+	linkOptions := options.Generic{
```
I took a deep look into this options package and am a little worried about pushing such a type all over. While there are some conveniences to this, I'd recommend retaining the `map[string]interface{}`. You could use something https://github.com/mitchellh/mapstructure to get the mapping to arbitrary structs.

The main reason this is problematic is [`GenerateFromModel `](https://github.com/docker/libnetwork/blob/master/pkg/options/options.go#L46). This is very tricky to get right and can lead to some serious head-scratchers.

mrjana replied:
@stevvooe This type is only used to transfer parameters opaquely from docker daemon to the driver without libnetwork needing to know anything about this. So it is used strictly for that purpose only and just to achieve backward compatibility. In the new world apis to libnetwork recognized objects will not use `options.Generic`.  In anycase,  `GenerateFromModel` is a less-ambitious api in that it just tries to convert a golang map[string]interface{} to a golang structure. Can you be more specific on what serious head-scratchers it can lead to?

stevvooe commented on daemon/execdriver/lxc/lxc_template_unit_test.go:
```diff
@@ -329,13 +317,8 @@ func TestCustomLxcConfigMiscOverride(t *testing.T) {
-				IPPrefixLen: 24,
-				Bridge:      "docker0",
-			},
+			Mtu:       1500,
```
`MTU`.

stevvooe commented on daemon/daemon.go:
```diff
@@ -15,6 +15,9 @@ import (
 
 	"github.com/docker/libcontainer/label"
+	"github.com/docker/libnetwork"
+	"github.com/docker/libnetwork/pkg/netlabel"
```
What is the difference between packages in `pkg` and other packages in libnetwork? Looking at the [godoc](http://godoc.org/github.com/docker/libnetwork), it seems these aren't getting indexed correctly.

mrjana replied:
packages in `pkg` are meant to be externally visible packages which can be consumed by anybody outside the context of libnetwork. Other packages in libnetwork are just internal packages to avoid putting all the libnetwork specific code in one directory because of the lack of any good way to put files of the same package in multiple directories

stevvooe replied:
@mrjana Why not use the `internal` package convention (https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit)? Should be coming with Go 1.5 but you may adopt the pattern now.

It is very odd to have a package declare that it has packages, when sub-directories of a package are already packages. These should just subdirectories.

mrjana replied:
@stevvooe I am aware of the upcoming internal proposal but it is going to be more than a convention. We will delay following that until it becomes a reality. For now, we will just remove `pkg`

mavenugo replied:
@stevvooe Thanks for the review. Regarding the `pkg/` structure, we wanted to follow the structure that `docker\docker` adopted : https://github.com/docker/docker/blob/master/pkg/README.md. The idea is that all of the packages under the `pkg` are utility packages used by libnetwork without specific to its internals and follows the same reasoning :
> Utility packages are kept separate from the docker core codebase to keep it as small and concise as possible. If some utilities grow larger and their APIs stabilize, they may be moved to their own repository under the Docker organization, to facilitate re-use by other projects. However that is not the priority.

Hence, we would like to keep that convention. But yes, we will fix the indexing issue that you brought up and that will make it much more clear.

mavenugo replied:
@stevvooe @mrjana godoc crawler is buggy and I fixed it for now and now it is indexing the directories properly : http://godoc.org/github.com/docker/libnetwork.
Yes, we can add good package documentation for all the packages. But that can be done later.

stevvooe commented on daemon/container.go:
```diff
@@ -582,60 +605,226 @@ func (container *Container) buildHostsFiles(IP string) error {
+		networkSettings.MacAddress = mac.(net.HardwareAddr).String()
+	}
+
+	if mapData, ok := epInfo[netlabel.PortMap]; ok {
```
Lose the unnecessary indent here:

```
if mapData, ok := epInfo[netlabel.PortMap]; !ok {
    return networkSettings, nil
}

// function meat!
```

mrjana replied:
Yes, that's the style we would like to follow as well. Will fix it.

stevvooe commented on daemon/container.go:
```diff
@@ -582,60 +605,226 @@ func (container *Container) buildHostsFiles(IP string) error {
+			for _, pp := range portMapping {
+				natPort := nat.NewPort(pp.Proto.String(), strconv.Itoa(int(pp.Port)))
+				natBndg := nat.PortBinding{HostIp: pp.HostIP.String(), HostPort: strconv.Itoa(int(pp.HostPort))}
+				if b, ok := networkSettings.Ports[natPort]; ok {
```
You don't actually need to test for presence before appending here. This can be replaced with the following:

```
networkSettings.Ports[natPort] = append(networkSettings.Ports[natPort], natBndg)
```

Append will do the right thing if the nil value is returned.

mrjana replied:
Agreed. This can be refactored.

LK4D4 commented:
@stevvooe Thanks for great review. We should invite you more often :)

ashahab-altiscale commented on daemon/execdriver/lxc/lxc_init_linux.go:
```diff
@@ -4,19 +4,38 @@ package lxc
+
+	// Make sure you setup the netns before you setup user since once you
+	// setup as a unprivileged used you cannot set into different namespace.
+	if args.NetNsPath != "" {
```
Have you tested this with LXC driver after dropping NET_ADMIN? LXC Caps are set in LXC(so by this time, they are already in place)

mrjana replied:
@ashahab-altiscale Thanks for the review. LXC driver seems to be failing after several rebases(even though I verified all the functionality manually when I made this change) but test-integration-cli was always passing in my machine(even now) so I was somewhat misled by that. I have fixed the issues to make LXC driver work again. Please take a look when I push the new diffs.

ashahab-altiscale replied:
Please make sure that you execdriver is lxc when running tests.
On May 16, 2015 8:12 AM, "Jana Radhakrishnan" <notifications@github.com>
wrote:

> In daemon/execdriver/lxc/lxc_init_linux.go
> <https://github.com/docker/docker/pull/13060#discussion_r30461812>:
>
> >  )
> >
> >  func finalizeNamespace(args *InitArgs) error {
> >  	if err := utils.CloseExecFrom(3); err != nil {
> >  		return err
> >  	}
> > +
> > +	// Make sure you setup the netns before you setup user since once you
> > +	// setup as a unprivileged used you cannot set into different namespace.
> > +	if args.NetNsPath != "" {
>
> @ashahab-altiscale <https://github.com/ashahab-altiscale> Thanks for the
> review. LXC driver seems to be failing after several rebases(even though I
> verified all the functionality manually when I made this change) but
> test-integration-cli was always passing in my machine(even now) so I was
> somewhat misled by that. I have fixed the issues to make LXC driver work
> again. Please take a look when I push the new diffs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13060/files#r30461812>.
>

mrjana replied:
Yeah it is lxc(I checked) according to the logs but it still succeeded. Do not know the answer for that.

LK4D4 replied:
Yeah, for me they passing too locally :/ Maybe apparmor or selinux problem. I think we need to figure this out.

ashahab-altiscale commented:
@jfrazelle Can you run an lxc-driver jenkins build for this?

unclejack commented:
@ashahab-altiscale The tests were run on LXC.

LK4D4 commented:
@ashahab-altiscale You can see, it falis on jenkins. There is red cross under all comments.

ashahab-altiscale commented:
Yup, looking.

mrjana commented:
@cpuguy83 @LK4D4 @calavera @stevvooe Update new diffs with almost all of the comments taken care of. 

@estesp I've reinstated all the test case variations for resolvconf update but now just considering the expected function alone. Please eyeball the test case and see if everything is fine.

@ashahab-altiscale I've re-implemented lxc integration in a different way so please take a look. And lxc is passing in jenkins with this implementation.

@ashahab-altiscale @LK4D4 the reason why the `test-integration-cli` for lxc was working but not jenkins was that jenkins was not running in a container while the locally run test suite was running inside a dind container.  `libnetwork` was setting up netns bind mounts in `/var/run/netns` and in the host they are mounted as `MS_SHARED` while inside a dind container they are not mounted as `MS_SHARED` and with the earlier design when I bind mounted `/var/run/netns/<ns>`, lxc umounted it which propogated to the host(because of `MS_SHARED`) and so setting into it inside the lxc init code started failing. Earlier I used to setup netns bind mounts inside /var/lib/docker/netns and that's why all my manual tests worked as expected and because of a design review comment I rightfully switched to `/var/run/netns` and didn't test lxc manually and only relied on the integration test suite and it was passing. So the mystery is resolved.

LK4D4 commented:
Yup, and `/var/run/netns` was pretty bad place anyway, because of `iproute2`.

mrjana commented:
@LKD4 The path is specified in `libnetwork`. May be it's more appropriate to have it in `/var/run/docker` but the reason I chose `/var/run/netns` is that this gives us a way to use `iproute2` to all the libnetwork created namespace as a debugging tool and we don't need to create another tool.

LK4D4 commented:
@mrjana Yes, but iproute2 is separate tool and it can be used for other purposes. `nsenter` tool can do pretty the same, but allow you specify file anywhere.
Okay, so where is namespaces now? And could you point to code line where it is specified?

mrjana commented:
@LK4D4 It's still at `/var/run/netns`. The code is here https://github.com/docker/libnetwork/blob/master/sandbox/namespace_linux.go#L15

LK4D4 commented:
I wonder why it isn't configurable. And I'm against placing it there.

mrjana commented:
@LK4D4 The whole point is this is part of the `sandbox` api so that the caller doesn't need to worry about  the OS specifics of how the sandbox is created and maintained to make the caller OS-agnostic and only return a sandbox key to uniquely identify the sandbox. So for linux the sandbox key happens to be the namespace path itself and for windows it would be a UUID string(from an early conversation with windows folks that's how they identify their network "compartments").

LK4D4 commented:
@mrjana Portability is hardly a reason for hardcoding behavior for all platforms in library. But okay, it's not a docker problem.
But I still don't like `/var/run/netns`, it is other tool namespace and we shouldn't spoil it with our namespaces.

mrjana commented:
@LK4D4 ok since as you suggested we can use `nsenter` for debugging purposes we should be able to use `/var/run/docker`. I will make that change.

LK4D4 commented:
`--userland-proxy=false` doesn't working, docker still spawning `docker-proxy` processes.

icecrime commented:
As far as I can tell, we have:

- Blockers

     - [ ] Zombie process
     - [ ] `--userland-proxy` not backported to libnetwork

- Things that could be improved

     - The `/var/run/netns`storage path should be left to the caller (and changed by Docker)
     - We're not super comfortable with the `Setns`/`defer` code paths, but @crosbymichael we just go with the changes and stress test the daemon

crosbymichael commented:
performance and runtime stuff:

LGTM

nerdalert commented:
I have been going through validating existing default networking flags in various mix and match combinations. I wrote it up in the form of a tutorial to keep it from looking like a jumbled mess of output. So far I haven't hit anything with the following OPTS:
```
-bridge=
--bip=
--fixed-cidr 
--ip=
--ipv6=
--mtu=
--dns=
--dns-search=
--hostname=
--net=
--mac-address=
```
Posting it to add extra data and maybe save time in reviews for folks it is a WIP so it will get some more updates tonight and tomorrow with more scenarios:
[libnetwork parity testing](https://gist.github.com/nerdalert/055f0562c90bd0796172)

mavenugo commented:
Thanks @nerdalert. This is helpful in addition to the automated integration tests & stress tests (thanks to @crosbymichael).

mrjana commented:
@LK4D4 @icecrime @crosbymichael Pushed new diffs with PR comments taken care of. More importantly fixed the zombie issue and backported making userland-proxy optional. I ran the whole `test-integration-cli` with `--userland-proxy=false` and is passing. Also made sure that proxy is not running when` --userland-proxy` is false.

icecrime commented on daemon/execdriver/lxc/driver.go:
```diff
@@ -214,10 +256,12 @@ func (d *driver) Run(c *execdriver.Command, pipes *execdriver.Pipes, startCallba
 	c.ProcessConfig.Args = append([]string{name}, arg...)
 
 	if err := createDeviceNodes(c.Rootfs, c.AutoCreatedDevices); err != nil {
+		killNetNsProc(proc)
```
It's cool that we now wait, but I find it unfortunate that it's called in each code path rather than `defer`ed: is there a reason not to?

mrjana replied:
Because this function does not return until the container has exited. I don't want to keep that netns proc until that. I just want to kill it once we positively know that the lxc has started the container.

crosbymichael replied:
In go functions will not return until all `defer`s are run

mrjana replied:
@crosbymichael I know, but in this function we wait on the waitLock right here:

```
        <-waitLock
        exitCode := getExitCode(c)
```

So you cannot put the `killNetNsProc` inside a defer

LK4D4 commented:
need rebase :/

LK4D4 commented:
@mrjana I'm not sure, but seems like your commits in wrong order now.

LK4D4 commented:
I found panic:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x30 pc=0x75206f]

goroutine 107 [running]:
github.com/docker/libnetwork.(*endpoint).Leave(0xc2082bcf30, 0xc2084753c0, 0x40, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:357 +0x2cf
github.com/docker/docker/daemon.(*Container).ReleaseNetwork(0xc208310000)
        /go/src/github.com/docker/docker/daemon/container.go:961 +0x45f
github.com/docker/docker/daemon.(*Container).cleanup(0xc208310000)
        /go/src/github.com/docker/docker/daemon/container.go:975 +0x36
github.com/docker/docker/daemon.(*containerMonitor).Close(0xc2081557a0, 0x0, 0x0)
        /go/src/github.com/docker/docker/daemon/monitor.go:86 +0x4a
github.com/docker/docker/daemon.func·031()
        /go/src/github.com/docker/docker/daemon/monitor.go:117 +0xbb
github.com/docker/docker/daemon.(*containerMonitor).Start(0xc2081557a0, 0x0, 0x0)
        /go/src/github.com/docker/docker/daemon/monitor.go:180 +0x7af
github.com/docker/docker/daemon.*containerMonitor.Start·fm(0x0, 0x0)
        /go/src/github.com/docker/docker/daemon/container.go:1584 +0x39
github.com/docker/docker/pkg/promise.func·001()
        /go/src/github.com/docker/docker/pkg/promise/promise.go:8 +0x2f
created by github.com/docker/docker/pkg/promise.Go
        /go/src/github.com/docker/docker/pkg/promise/promise.go:9 +0xfb
```
I created two containers with `--net=host` and then leave from both. Also you probably want to write test for this situation :)

LK4D4 commented:
After daemon crash I can't start old container with:
```
Error response from daemon: Cannot start container c6732738abba: open /var/run/docker/netns/c6732738abba: permission denied
Error: failed to start containers: [c6732738abba]
```
In logs:
```
ERRO[0023] leaving endpoint failed: no container attached to the endpoint
```

LK4D4 commented:
Last error probably was here forever, but now it became little more severe. I think we need to call `container.cleanup()` in `daemon.register` instead of just `container.Unmount`

icecrime commented:
On the good side of things, tests are green with `--userland-proxy=false`.

estesp commented:
@mrjana thanks for the `resolv.conf` test update--it looks good.  Based on that, it appears that any kind of live watching of the host resolver config is now replaced from an implementation perspective by DNS setup step during container start?  Given we had no way to do live update of running containers, this seems fully reasonable and at functional parity with today's implementation.

mrjana commented:
@estesp Yes, that was the only design change which functionally doesn't change anything.

mrjana commented:
@LK4D4 Thanks for catching the panic issue. Pushed a new set of diffs with the fix for that a test case for that in libnetwork as well. Also rebased.

/cc @icecrime @crosbymichael

LK4D4 commented:
```
Error response from daemon: Cannot start container c6732738abba: open /var/run/docker/netns/c6732738abba: permission denied
Error: failed to start containers: [c6732738abba]
```
still here.
And I don't see that you added some code to fix it.
I reproduced like
```
docker run -d busybox sleep 1000
kill docker with 4 ctrl-C
start docker
docker start old container
```

LK4D4 commented:
also somehow you need rebase again :/

mrjana commented:
@LK4D4 I pushed new diffs which attempts to fix the issue that you are having. BTW, I was not able to reproduce this issue at all in my local machine. Probably something to do with the version of kernel you are running. 

In any case the way I've fixed it is to to remove the namespace path file if it exists during a new join for an old container and recreate the namespace for the container from scratch. Please test it in your machine and see if it works ok.

This is rebased again.

jfrazelle commented:
testing now on 4.0.4

On Tue, May 19, 2015 at 5:03 PM, Jana Radhakrishnan <
notifications@github.com> wrote:

> @LK4D4 <https://github.com/LK4D4> I pushed new diffs which attempts to
> fix the issue that you are having. BTW, I was not able to reproduce this
> issue at all in my local machine. Probably something to do with the version
> of kernel you are running.
>
> In any case the way I've fixed it is to to remove the namespace path file
> if it exists during a new join for an old container and recreate the
> namespace for the container from scratch. Please test it in your machine
> and see if it works ok.
>
> This is rebased again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13060#issuecomment-103701507>.
>

mrjana commented:
Thanks @jfrazelle

jfrazelle commented:
LGTM

icecrime commented:
LGTM

mavenugo commented:
Thanks @mrjana @icecrime @LK4D4 @crosbymichael @jfrazelle @estesp @ashahab-altiscale @stevvooe @cpuguy83 for a good review & merge.

:tada: :balloon: :fireworks:

fgbreel commented:
Finally! :+1:

noizu commented:
Is there a rough estimate of when this extension may, if approved,make it into the stable branch? 6 months, 12 months, etc. 

This perfectly answers some scenarios I have but I need to decide if I implement custom work arounds and leverage third party tools or work on other components of the system until this is in place or those other work arounds need to be done.  

Excellent work meanwhile. Simplified elastic-ip like routing between containers is definitely a major hole in the existing feature set.

stevvooe commented:
@noizu This isn't really the right venue to ask this sort of question. I suggest that you use either the mailing list (docker-dev@googlegroups.org) or IRC #docker-dev.
